From 50a99a71f91515e8b984e3fe8f9bd1dd44ec6c7a Mon Sep 17 00:00:00 2001
From: "Druzga, LukaszX" <lukaszx.druzga@intel.com>
Date: Mon, 20 Jul 2020 11:01:52 +0200
Subject: [PATCH] PONRTSYS-7673: xrx500-eth 18000000.ssx1_1:eth@a00000 eth0_0
 (uninitialized): Unable to find phydev

Add support for .reset callback in xgmac_mdio driver.
Support a flexible list of gpios to be defined in devicetree,
if there are multiple external phys.
---
 .../ethernet/lantiq/switch-api/mac/xgmac_mdio.c    | 33 +++++++++++++++++++++-
 1 file changed, 32 insertions(+), 1 deletion(-)

diff --git a/drivers/net/ethernet/lantiq/switch-api/mac/xgmac_mdio.c b/drivers/net/ethernet/lantiq/switch-api/mac/xgmac_mdio.c
index e1727929e646..b2ffb2865402 100644
--- a/drivers/net/ethernet/lantiq/switch-api/mac/xgmac_mdio.c
+++ b/drivers/net/ethernet/lantiq/switch-api/mac/xgmac_mdio.c
@@ -43,6 +43,7 @@
 #include <xgmac_mdio.h>
 #include <xgmac.h>
 #ifdef __KERNEL__
+#include <linux/gpio/consumer.h>
 #include <linux/of_mdio.h>
 #endif
 
@@ -489,6 +490,36 @@ static int xgmac_mdio_write(struct mii_bus *bus, int phyadr, int phyreg,
 	return ret;
 }
 
+int xgmac_mdio_reset(struct mii_bus *bus)
+{
+	struct device *dev = bus->parent;
+	struct gpio_descs *descs = NULL;
+	int i;
+
+	if (!dev)
+		return 0;
+
+	descs = gpiod_get_array_optional(dev, "reset", GPIOD_OUT_LOW);
+	if (!descs)
+		return 0;
+
+	/* set reset gpios as output and activate reset */
+	for (i = 0; i < descs->ndescs; ++i)
+		gpiod_direction_output(descs->desc[i], 1);
+
+	msleep(100);
+
+	/* deactivate reset gpios */
+	for (i = 0; i < descs->ndescs; ++i)
+		gpiod_set_value(descs->desc[i], 0);
+
+	msleep(100);
+
+	gpiod_put_array(descs);
+
+	return 0;
+}
+
 /* API to register mdio.
  */
 int xgmac_mdio_register(void *pdev)
@@ -518,7 +549,7 @@ int xgmac_mdio_register(void *pdev)
 	new_bus->name = "xgmac_phy";
 	new_bus->read = xgmac_mdio_read;
 	new_bus->write = xgmac_mdio_write;
-	new_bus->reset = NULL;
+	new_bus->reset = xgmac_mdio_reset;
 	snprintf(new_bus->id, MII_BUS_ID_SIZE, "%s-%x", new_bus->name,
 		 pdata->mac_idx);
 	new_bus->priv = pdev;
